src/deroff.rs

```diff
@@ -296,6 +328,27 @@ impl Deroffer {
             c => c.chars().all(|c| c.is_whitespace()),
         }
     }
+
+    fn not_whitespace(s: &str, idx: usize) -> bool {
+        //         # Note that this return False for the empty string (idx >= len(self.s))
+        //         ch = self.s[idx:idx+1]
+        //         return ch not in ' \t\n'
+        let ch = s
```

Comment:
Could maybe succinctify the function body a little:
```rust
        s.get(idx..(idx + 1))
            .map(|string| string == " " || string == "\t" || string == "\n")
            .unwrap_or_default()
```

Reply:
My suggestion
```rust
        s.get(idx..(idx + 1))
            .map(|s| s.chars().next())
            .map_or(false, |ch| ch == ' ' || ch == '\t' || ch == '\n')
```

Reply:
Thanks for commenting!

> My suggestion
> 
> ```rust
>         s.get(idx..(idx + 1))
>             .map(|s| s.chars().next())
> ```

In that line there you'd want to use `and_then` instead of `map`, since `next` is going to return and `Option<char>` which will be wrapped in `Some` by `map` to be `Option<Option<char>>`. Could also `unwrap` since the indexing restrictions in `get` ensure that we _have_ to have a single `char` in `s`. I also rely on that in my version to allow `"...".contains(...)` to work—it only does what we want because if `get(idx..(idx + 1))` returns a `Some`, it must be a valid single-byte (and so also single-char) UTF-8 string.

> ```rust
>             .map_or(false, |ch| ch == ' ' || ch == '\t' || ch == '\n')
> ```

I'm totally cheating in my suggestion, but I don't know whether there's a performance penalty for that cheat. It just lets me avoid the longer combination of individual equality checks, but isn't necessarily performant.

If we didn't want to cheat by using `str::contains`, I'd likely go with

```rust
        s.get(idx..(idx + 1))
            .map(|string| string == " " || string == "\t" || string == "\n")
            .unwrap_or_default()
```

Up to @PureW whether he wants to modify his original.

Reply:
Nice discussion! I went with the first suggestion as it seemed the easiest to read after the fact.